File: lang/java/src/fun/vac/algorithms/number_theory/LeastCommonMultiple.java
package fun.vac.algorithms.number_theory;

import fun.vac.algorithms.number_theory.GreatestCommonDivisor;

public class LeastCommonMultiple {

    private LeastCommonMultiple() {}

    public static long reduceToBinaryGCD(long m, long n) {
        if (m < 0) {
            m = -m;
        }
        if (n < 0) {
            n = -n;
        }

        if (m == 0 || n == 0) {
            return 0;
        } else {
            return m / GreatestCommonDivisor.iterativeBinaryGCD(m, n) * n;
        }
    }

    public static long reduceToEuclidean(long m, long n) {
        if (m < 0) {
            m = -m;
        }
        if (n < 0) {
            n = -n;
        }

        if (m == 0 || n == 0) {
            return 0;
        } else {
            return m / GreatestCommonDivisor.iterativeEuclidean(m, n) * n;
        }
    }
}

File: lang/python/src/fun/vac/algorithms/number_theory/LeastCommonMultiple.py
from fun.vac.algorithms.number_theory import GreatestCommonDivisor

def reduceToBinaryGCD(m: int, n: int) -> int:
    if m < 0:
        m = -m
    if n < 0:
        n = -n

    if m == 0 or n == 0:
        return 0
    else:
        return m // GreatestCommonDivisor.iterativeBinaryGCD(m, n) * n

def reduceToEuclidean(m: int, n: int) -> int:
    if m < 0:
        m = -m
    if n < 0:
        n = -n

    if m == 0 or n == 0:
        return 0
    else:
        return m // GreatestCommonDivisor.iterativeEuclidean(m, n) * n
